### ISSUE TITLE
3.0.0 Code Base Refactoring

### ISSUE THREAD

**Issue by LordFokas** [NanoHttpd/nanohttpd#278]
Alright, so, soon enough I'll have enough time to look into refactoring the code base to make it easier to use and understand, so there's a few things I'd like to discuss first so that we all agree to what's going to happen.

Since we're going to make a few breaking changes, might as well take the chance and make ALL the breaking changes and let it settle.

I'm ok with anyone providing input, but I mostly want to pull @ritchieGitHub and @elonen into this.

So, here goes:

**I) Splitting The Project**
Currently NanoHTTPD has two kinds of code: library (NanoHTTPD / NanoWSD / NanoSSED in the future) and servers/misc/utilities (file server, nanolets server, etc).
My idea is to split the library modules into a separate repository and then import it as a git submodule.
This would make things easier to manage. Splitting the Java 1.1 version from a branch into another repository might also help.

The general idea is to split the lib out and make it so that every "final result" you may want to use is a master branch in a repository.

**II) Breaking Files Apart**
While having everything in a single file helps pull the code into another project, it only really works for the HTTP server as everything else has other dependencies and therefore doesn't work by importing a single file. That said I would like to break those apart (at least the protocol servers) and extract most inner types to their own files. This not only is what makes sense but it also makes managing code and debugging a lot easier. To be honest so much bureaucracy has been added around the project in the past few years that I don't think having to download a whole package instead of a file would bother anyone or make it any more difficult to integrate in any other project.

**III) Protocol Server Plugins**
This one is simple. Handling HTTP / WS / SSE  requests (or sockets) is now delegated to an interface that can be plugged into the server. If the implementor is null a default error response is sent (404/close socket/etc).

**IV) Server Constructor Refactoring**
Protocol servers (HTTP/WS/SSE) should have a constructor that receives configurations, and a default (empty) constructor that runs with the default configs (port 80, no domain name, etc).

Misc servers (nanolets, fileserver) should have an empty constructor where they create the default server (HTTP) through the default configs (empty constructor) and set themselves as the HTTP request handler; a second constructor where they receive a set of configurations and construct a new default server and pass it the configurations; and a third constructor where they receive any server and set themselves as the HTTP handler.

This allows for extra flexibility in servers (for example, having a WS server that also serves files on HTTP would be trivial). A Configuration class / hierarchy is probably the best approach for this.

**In conclusion...**
These are the breaking changes that I'd like to implement to make everything better. Most of them are independent of each other and can be harmlessly discarded. The point of this issue is to get some consent about what to do and what not to do. That said, I hereby explicitly request your opinions on the four points above :)

**Comment by ritchieGitHub** [comment]
OK my thoughts about the subjects:
**I) Splitting The Project**
here I am more strongly opposed, the maven modularisation is good enough. Specially as it would complicate the releasing a whole lot. Having said that, there are some things that we could do without to much overhead. I think the modules are to tightly bound for manageable separate repositories. That would introduce a lot of verification tasks or jobs to keep the modules from "getting old", as we are all doing nano for a few hours a week .... I think that's a bad idea, if we had more resources I could have another opinion.
 
- move the 1.1 version in a own repository is good (the other branches where old and are now deleted) 
- put the core in one repository and the rest in another is a possibility. That would keep release/maven work in range. But still I think it is not worth the troule.

**II) Breaking Files Apart**
I have no opposition to this proposal, even splitting the core would a good thing, when the extendibility is improved massively. 
The only problem is that we need a new definition of "nano":
- size of the jar?
- total lines of code?
- some other metric
Important is that the metric can be calculated automatically so we do not lose the perspective of staying nano. Because that's the main point of nanohttpd.

**III) Protocol Server Plugins**
yes yes yes that is what we need ;-)

**IV) Server Constructor Refactoring**
Yes, but ;-) 
I would create only one constructor in the core and plugins that take a "builder" as a parameter that can be configured using a fluent api. The defaults can be in the builder and the main code would not be "polluted" with multiple constructors.

**conclusion**
I would let "I" be for the mean time, and concentrate on the rest. I will try to find a way around the management overhead of using multiple repositories. (Maven/release/travis/site ....)

**Comment by ritchieGitHub** [comment]
@LordFokas and @elonen the 1.1 branch is move to a own repo.

**Comment by elonen** [comment]
My original reasons for creating NanoHTTPD were to:

1. Make it dead simple to write a minimalistic HTTP enabled program. Copy the single .java file, override serve() and be done with it.
2. Make it easy to grasp. The less modules, public classes/function and lines of code as possible, the easier it is to read through, see what's going on, and understand how to use it.
3. Keep it Java 1.1 to support early Java enabled mobile devices and incomplete JDKs on libre operating systems.

Clearly, reason 3 is now obsolete - moving it to a separate repo is a great idea.

The installation part ("just copy the single .java file") of reason 1 has been, depending on one's viewpoint, either fixed or destroyed by Maven and Jar based releases. (Anyway, to repeat: I haven't done any serious Java programming for at least 10 years, so I'm probably not the best source of insight on contemporary Java deployment.)

The integration part of 1 ("override serve() and be done with it") is still valid goal I guess.

Reason 2 is in constant threat from the pressure to add more features (like full HTTP 1.1 compliance). An incredibly smart refactoring (aiming for minimizing lines to read) and pluginization of non-essential features would help a lot in this respect. It's very hard of course, as large feature set, extensibility, backwards compatibility and easy-to-grasp codebase a fundamentally conflicting goals.

The rise of Android may have introduced new reasons for people using NanoHTTPD. Not sure what they are though, if any. Small binaries, peharps? Also, some people seem to be using it more or less as a standalone server instead of as base for custom apps. I don't really know. The market research department hasn't been very effective. ;)

**Comment by ritchieGitHub** [comment]
@LordFokas 2.3.0 released, so you are free to go ;-)

**Comment by LordFokas** [comment]
My semester started this week and I'm still running a full time job until next week so time is a bit tight here for a while... I'll try to keep you updated.

**Comment by ritchieGitHub** [comment]
@elonen  and @LordFokas 
I have to stop developing for a few weeks, because I suffer from a tragic loss in my family....
I will be back, but now I am mourning for a while ....

**Comment by elonen** [comment]
Thank you for letting us know.

-Jarno

**Comment by ritchieGitHub** [comment]
@LordFokas @elonen I am  back.

 @LordFokas could you send me your email adress? you can find mine in the ".travis.yml" file ;-)

**Comment by trnchawla** [comment]
Please tell me the latest branch to pull.

**Comment by sudarh** [comment]
not sure anybody is still watching this trail, but thought of posting. @LordFokas , what I understand is that SSE actually runs on HTTP. The way it is differentiated is using the "Accept" header. that is set as text/event-stream. So, ideally , NanoHTTP should support SSE internally and have a different router to it. In that sense, I think the process() in routerNanohttpd ( if you are using that) can be extended to supported SSE. So, when you say " Protocol servers (HTTP/WS/SSE) should have a constructor that receives configurations, and a default (empty) constructor that runs with the default configs (port 80, no domain name, etc)"  I am not sure I get that. A server can support regular HTTP as well as  SSE, it is just the HTTP header that indicates that.

**Comment by sudarh** [comment]
The way I was thinking of supporting SSE is to extend RouterNanoHTTPD and override process() and call a different handler for SSE requests. The handler will not be closed till server requests or a client closes it or it results in a broken pipe. The SSE handler will take care of parsing the EventSource parameters such as lastEventId etc.  Please let me know if this make sense or not
